### PR TITLE
[Core] Added number extension

### PR DESCRIPTION
**Pull request by krsriq**
### What is the reason for this PR?

Number extension is not yet in core.

- [x] A new feature

### Author's checklist

- [x] Follow the [Contribution Guide](https://github.com/FakerPHP/Faker/blob/main/.github/CONTRIBUTING.md)
- [x] New features and changes are [documented](https://github.com/FakerPHP/fakerphp.github.io)

### Summary of changes

Added the number extension to the core.

Should we move the number methods to a helper class (as static methods), so it's easier to use them in other (internal) methods, such as for #256?

### Review checklist

- [ ] All checks have passed
- [ ] Changes are approved by maintainer

### CODE REVIEW

File: src/Faker/Core/Number.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+{
+    /**
+     * Returns a random number between $int1 and $int2 (any order)
+     *
+     * @param int $int1 default to 0
+     * @param int $int2 defaults to 32 bit max integer, ie 2147483647
+     *
+     * @example 79907610
+     */
+    public function numberBetween($int1 = 0, $int2 = 2147483647): int
```
Comment (localheinz) [inline src/Faker/Core/Number.php:22]:
```suggestion
    public function numberBetween(int $int1 = 0, int $int2 = 2147483647): int
```

File: src/Faker/Core/Number.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        return mt_rand(0, 9);
+    }
+
+    /**
+     * Generates a random digit, which cannot be $except
+     *
+     * @param int $except
+     */
+    public function randomDigitNot($except): int
```
Comment (localheinz) [inline src/Faker/Core/Number.php:43]:
```suggestion
    public function randomDigitNot(int $except): int
```

File: src/Faker/Core/Number.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Return a random float number
+     *
+     * @param int|null  $nbMaxDecimals
+     * @param float|int $min
+     * @param float|int $max
+     *
+     * @example 48.8932
+     */
+    public function randomFloat($nbMaxDecimals = null, $min = 0, $max = null): float
```
Comment (localheinz) [inline src/Faker/Core/Number.php:71]:
```suggestion
    public function randomFloat(int $nbMaxDecimals = null, float $min = 0, float $max = null): float
```

File: src/Faker/Core/Number.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+     * Returns a random integer with 0 to $nbDigits digits.
+     *
+     * The maximum value returned is mt_getrandmax()
+     *
+     * @param int|null $nbDigits Defaults to a random number between 1 and 9
+     * @param bool     $strict   Whether the returned number should have exactly $nbDigits
+     *
+     * @example 79907610
+     */
+    public function randomNumber($nbDigits = null, $strict = false): int
```
Comment (localheinz) [inline src/Faker/Core/Number.php:104]:
```suggestion
    public function randomNumber(int $nbDigits = null, bool $strict = false): int
```

File: src/Faker/Extension/NumberExtension.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * Returns a random number between $int1 and $int2 (any order)
+     *
+     * @param int $int1 default to 0
+     * @param int $int2 defaults to 32 bit max integer, ie 2147483647
+     *
+     * @example 79907610
+     */
+    public function numberBetween($int1, $int2): int;
```
Comment (localheinz) [inline src/Faker/Extension/NumberExtension.php:18]:
```suggestion
    public function numberBetween(int $int1, int $int2): int;
```

File: src/Faker/Extension/NumberExtension.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+     * Returns a random number between 0 and 9
+     */
+    public function randomDigit(): int;
+
+    /**
+     * Generates a random digit, which cannot be $except
+     *
+     * @param int $except
+     */
+    public function randomDigitNot($except): int;
```
Comment (localheinz) [inline src/Faker/Extension/NumberExtension.php:30]:
```suggestion
    public function randomDigitNot(int $except): int;
```

File: src/Faker/Extension/NumberExtension.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Return a random float number
+     *
+     * @param int|null  $nbMaxDecimals
+     * @param float|int $min
+     * @param float|int $max
+     *
+     * @example 48.8932
+     */
+    public function randomFloat($nbMaxDecimals, $min, $max): float;
```
Comment (localheinz) [inline src/Faker/Extension/NumberExtension.php:46]:
```suggestion
    public function randomFloat(int $nbMaxDecimals, float $min, float $max): float;
```

File: src/Faker/Core/Number.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+final class Number implements Extension\NumberExtension
+{
+    /**
+     * Returns a random number between $int1 and $int2 (any order)
+     *
+     * @param int $int1 default to 0
+     * @param int $int2 defaults to 32 bit max integer, ie 2147483647
+     *
+     * @example 79907610
+     */
```
Comment (localheinz) [inline src/Faker/Core/Number.php:21]: What do you think about removing the DocBlocks here?

It is automatically inherited from the interface that this class implements.

File: src/Faker/Core/Number.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+     * @param int|null $nbDigits Defaults to a random number between 1 and 9
+     * @param bool     $strict   Whether the returned number should have exactly $nbDigits
+     *
+     * @example 79907610
+     */
+    public function randomNumber($nbDigits = null, $strict = false): int
+    {
+        if (!is_bool($strict)) {
+            throw new \InvalidArgumentException('randomNumber() generates numbers of fixed width. To generate numbers between two boundaries, use numberBetween() instead.');
+        }
```
Comment (localheinz) [inline src/Faker/Core/Number.php:108]: Can we remove this when we add the type declaration?

File: src/Faker/Core/Number.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+        if (null === $nbDigits) {
+            $nbDigits = $this->randomDigitNotNull();
+        }
+        $max = 10 ** $nbDigits - 1;
+
+        if ($max > mt_getrandmax()) {
+            throw new \InvalidArgumentException('randomNumber() can only generate numbers up to mt_getrandmax()');
+        }
```
Comment (localheinz) [inline src/Faker/Core/Number.php:117]: Can we remove this when we add the type declaration?

File: src/Faker/Generator.php [hunk 11]
```diff
@@ -365,6 +362,36 @@ public function isbn13(): string
         return $this->ext(Extension\BarcodeExtension::class)->isbn13();
     }
 
+    public function numberBetween($int1 = 0, $int2 = 2147483647): int
+    {
+        return $this->ext(Extension\NumberExtension::class)->numberBetween($int1, $int2);
```
Comment (localheinz) [inline src/Faker/Generator.php:367]:
```suggestion
        return $this->ext(Extension\NumberExtension::class)->numberBetween((int) $int1, (int) $int2);
```

File: src/Faker/Generator.php [hunk 12]
```diff
@@ -365,6 +362,36 @@ public function isbn13(): string
...
+    }
+
+    public function randomDigit(): int
+    {
+        return $this->ext(Extension\NumberExtension::class)->randomDigit();
+    }
+
+    public function randomDigitNot($except): int
+    {
+        return $this->ext(Extension\NumberExtension::class)->randomDigitNot($except);
```
Comment (localheinz) [inline src/Faker/Generator.php:377]:
```suggestion
        return $this->ext(Extension\NumberExtension::class)->randomDigitNot((int) $except);
```

File: src/Faker/Generator.php [hunk 13]
```diff
@@ -365,6 +362,36 @@ public function isbn13(): string
...
+    }
+
+    public function randomDigitNotNull(): int
+    {
+        return $this->ext(Extension\NumberExtension::class)->randomDigitNotNull();
+    }
+
+    public function randomFloat($nbMaxDecimals = null, $min = 0, $max = null): float
```
Comment (localheinz) [inline src/Faker/Generator.php:385]: Can we add a DocBlock here?

File: src/Faker/Generator.php [hunk 14]
```diff
@@ -365,6 +362,36 @@ public function isbn13(): string
...
+    {
+        return $this->ext(Extension\NumberExtension::class)->randomFloat($nbMaxDecimals, $min, $max);
```
Comment (localheinz) [inline src/Faker/Generator.php:387]:
```suggestion
        return $this->ext(Extension\NumberExtension::class)->randomFloat((int) $nbMaxDecimals, (float) $min, (float) $max);
```

File: src/Faker/Generator.php [hunk 15]
```diff
@@ -365,6 +362,36 @@ public function isbn13(): string
...
+    }
+
+    public function randomNumber($nbDigits = null, $strict = false): int
+    {
+        return $this->ext(Extension\NumberExtension::class)->randomNumber($nbDigits, $strict);
```
Comment (localheinz) [inline src/Faker/Generator.php:392]:
```suggestion
        return $this->ext(Extension\NumberExtension::class)->randomNumber((int) $nbDigits, (bool) $strict);
```

File: src/Faker/Core/Number.php (hunk 1 above)
Comment (pimjansen) [reply src/Faker/Core/Number.php:22]: Not

File: src/Faker/Extension/NumberExtension.php (hunk 5 above)
Comment (pimjansen) [reply src/Faker/Extension/NumberExtension.php:18]: Same here is prev. Dont think int2 is a good name.

Min max maybe?

File: src/Faker/Core/Number.php (hunk 8 above)
Comment (krsriq) [reply src/Faker/Core/Number.php:21]: Removed the docblocks from this class now 👍️
I'll remove them from the other core extension classes (Barcode and File) in a separate PR.

File: src/Faker/Core/Number.php (hunk 1 above)
Comment (krsriq) [reply src/Faker/Core/Number.php]: @pimjansen I've renamed the arguments now, thanks!

File: src/Faker/Generator.php (hunk 13 above)
Comment (krsriq) [reply src/Faker/Generator.php:385]: I've added dockblocks to the generator ext methods now (will do so for the other extension methods in a separate PR).

**Comment by krsriq** [comment]
@localheinz @pimjansen - thanks, I've applied your suggestions now and fixed a couple of things. Could you please have another look?

File: src/Faker/Core/Number.php [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        $result = self::numberBetween(0, 8);
+
+        if ($result >= $except) {
+            ++$result;
+        }
+
+        return $result;
+    }
+
+    public function randomDigitNotNull(): int
```
Comment (pimjansen) [inline src/Faker/Core/Number.php:38]: This is a weird method name where i guess whereNotZero is meant?

Since the return type is never null

Comment (krsriq) [reply src/Faker/Core/Number.php]: You're right @pimjansen, I renamed the method.

**Review by Nyholm — APPROVED** [review]
Thank you

**Review by pimjansen — APPROVED** [review]

**Comment by krsriq** [comment]
Thanks!

**Pull request merged**